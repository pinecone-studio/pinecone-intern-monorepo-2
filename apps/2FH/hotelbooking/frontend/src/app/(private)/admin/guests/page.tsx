import GuestsPage from '@/components/guests/GuestsPage';

const page = () => {
  return (
    <div>
      <GuestsPage />
    </div>
  );
};

export default page;
