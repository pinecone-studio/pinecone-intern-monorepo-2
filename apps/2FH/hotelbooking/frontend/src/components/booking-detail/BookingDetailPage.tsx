'use client';

import React, { useState } from 'react';
import { ChevronLeft } from 'lucide-react';
import { useGetBookingQuery, useUpdateBookingMutation, useHotelQuery, BookingStatus } from '@/generated';
import { useRouter } from 'next/navigation';
import { CancellationRules } from './components/CancellationRules';
import { BookingInfo } from './components/BookingInfo';
import { PropertySupport } from './components/PropertySupport';
import { HotelImage } from './components/HotelImage';
import { CancelBookingModal } from './components/CancelBookingModal';
import { HotelLoader } from '../loadingComponent/Loader';

/* eslint-disable-next-line */
export const BookingDetailPageUser = ({ bookingId }: { bookingId: string }) => {
  const router = useRouter();
  const [isCancelling, setIsCancelling] = useState(false);
  const [showCancelModal, setShowCancelModal] = useState(false);

  const { data: bookingData, loading: bookingLoading } = useGetBookingQuery({
    variables: { getBookingId: bookingId },
  });

  const [updateBooking] = useUpdateBookingMutation();

  const booking = bookingData?.getBooking;

  const { data: hotelData, loading: hotelLoading } = useHotelQuery({
    skip: !booking?.hotelId,
    variables: { hotelId: booking?.hotelId || '' },
  });

  const hotelNumber = hotelData?.hotel?.phone;

  const handleCancelBooking = async () => {
    if (!booking) return;

    setIsCancelling(true);
    try {
      await updateBooking({
        variables: {
          updateBookingId: booking.id,
          input: {
            status: BookingStatus.Cancelled,
          },
        },
      });
      setShowCancelModal(false);
      window.location.reload();
    } catch (error) {
      console.error('Error cancelling booking:', error);
    } finally {
      setIsCancelling(false);
    }
  };

  if (bookingLoading || hotelLoading) {
    return (
      <div className="flex items-center justify-center h-screen">
        <HotelLoader />
      </div>
    );
  }

  if (!booking || !hotelData?.hotel) {
    return (
      <div className="min-h-screen bg-gray-50 flex items-center justify-center">
        <div className="text-center">
          <div className="text-lg font-semibold mb-2">No Booking Found</div>
          <div className="text-gray-600">The requested booking could not be found.</div>
        </div>
      </div>
    );
  }

  const hotel = hotelData.hotel;
  const hotelImage = hotel.images?.[0] || '/Images/NoImage.png';
  const isBooked = booking.status === BookingStatus.Booked;

  return (
    <div className="min-h-screen bg-gray-50">
      <div className="bg-white border-b border-gray-200 px-4 py-3">
        <div className="flex items-center gap-3">
          <button onClick={() => router.back()} className="p-2 hover:bg-gray-100 rounded-full transition-colors">
            <ChevronLeft className="w-5 h-5" />
          </button>
          <h1 className="text-lg font-semibold">Booking Details</h1>
        </div>
      </div>

      <div className="max-w-6xl mx-auto p-4">
        <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
          <div className="space-y-6">
            <CancellationRules />
            <BookingInfo
              hotelName={hotel.name}
              bookingStatus={booking.status!}
              checkInDate={booking.checkInDate}
              checkOutDate={booking.checkOutDate}
              adults={booking.adults || 0}
              childrens={booking.children || 0}
              roomId={booking.roomId}
              isBooked={isBooked}
              onCancelClick={() => setShowCancelModal(true)}
            />
            <PropertySupport hotelName={hotel.name} bookingId={booking.id} hotelNumber={hotelNumber || ''} />
          </div>

          {/* Right Column - Hotel Image and Info */}
          <div className="space-y-6">
            <HotelImage hotelName={hotel.name} hotelLocation={hotel.location} hotelImage={hotelImage} hotelRating={hotel.rating} />
          </div>
        </div>
      </div>

      <CancelBookingModal isOpen={showCancelModal} isCancelling={isCancelling} onClose={() => setShowCancelModal(false)} onConfirm={handleCancelBooking} />
    </div>
  );
};
