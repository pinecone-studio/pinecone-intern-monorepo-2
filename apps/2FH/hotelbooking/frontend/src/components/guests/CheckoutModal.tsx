'use client';
import React, { useState } from 'react';
import { useUpdateBookingMutation, BookingStatus, Response } from '@/generated';
import { Dialog, DialogContent, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';
import { RadioGroup, RadioGroupItem } from '@/components/ui/radio-group';
import { Label } from '@/components/ui/label';
import { Loader2 } from 'lucide-react';

interface CheckoutModalProps {
  bookingId: string;
  currentStatus: string;
  onStatusUpdate?: () => void;
}

const CheckoutModal = ({ bookingId, currentStatus, onStatusUpdate }: CheckoutModalProps) => {
  const [open, setOpen] = useState(false);
  const [selectedStatus, setSelectedStatus] = useState(currentStatus);
  const [error, setError] = useState<string | null>(null);
  const [updateBooking, { loading }] = useUpdateBookingMutation();

  const statusOptions = [
    { value: BookingStatus.Booked, label: 'Booked', color: 'bg-blue-600 text-white' },
    { value: BookingStatus.Completed, label: 'Completed', color: 'bg-green-600 text-white' },
    { value: BookingStatus.Cancelled, label: 'Cancelled', color: 'bg-orange-500 text-white' },
  ];

  const handleStatusUpdate = async () => {
    try {
      setError(null);
      const result = await updateBooking({
        variables: {
          updateBookingId: bookingId,
          input: {
            status: selectedStatus as BookingStatus,
          },
        },
      });

      if (result.data?.updateBooking === Response.Success) {
        setOpen(false);
        onStatusUpdate?.();
      } else {
        setError('Failed to update booking status. Please try again.');
      }
    } catch (error) {
      console.error('Error updating booking status:', error);
      setError('An error occurred while updating the booking status. Please try again.');
    }
  };

  const handleOpenChange = (newOpen: boolean) => {
    setOpen(newOpen);
    if (newOpen) {
      setError(null);
    }
  };

  return (
    <Dialog open={open} onOpenChange={handleOpenChange}>
      <DialogTrigger asChild>
        <Button className="bg-blue-600 hover:bg-blue-700 text-white">Change Status</Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-md">
        <DialogHeader>
          <DialogTitle>Update Booking Status</DialogTitle>
        </DialogHeader>

        <div className="space-y-6">
          <div>
            <Label className="text-sm font-medium text-gray-600 mb-3 block">Current Status</Label>
            <Badge className={statusOptions.find((s) => s.value === currentStatus)?.color || 'bg-gray-600 text-white'}>{currentStatus.charAt(0) + currentStatus.slice(1).toLowerCase()}</Badge>
          </div>

          <div>
            <Label className="text-sm font-medium text-gray-600 mb-3 block">Select New Status</Label>
            <RadioGroup value={selectedStatus} onValueChange={setSelectedStatus} className="space-y-3">
              {statusOptions.map((status) => (
                <div key={status.value} className="flex items-center space-x-2">
                  <RadioGroupItem value={status.value} id={status.value} />
                  <Label htmlFor={status.value} className="flex items-center gap-2 cursor-pointer">
                    <Badge className={status.color}>{status.label}</Badge>
                  </Label>
                </div>
              ))}
            </RadioGroup>
          </div>

          {error && (
            <div className="bg-red-50 border border-red-200 rounded-md p-3">
              <p className="text-sm text-red-600">{error}</p>
            </div>
          )}

          <div className="flex justify-end gap-3">
            <Button variant="outline" onClick={() => setOpen(false)} disabled={loading}>
              Cancel
            </Button>
            <Button onClick={handleStatusUpdate} disabled={loading || selectedStatus === currentStatus} className="bg-blue-600 hover:bg-blue-700 text-white">
              {loading && <Loader2 className="mr-2 h-4 w-4 animate-spin" />}
              Update Status
            </Button>
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
};

export default CheckoutModal;
