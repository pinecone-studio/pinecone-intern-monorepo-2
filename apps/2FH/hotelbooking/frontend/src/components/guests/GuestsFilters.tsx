'use client';
import React from 'react';
import { Input } from '@/components/ui/input';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '@/components/ui/select';
import { Search } from 'lucide-react';

interface GuestsFiltersProps {
  searchTerm: string;
  statusFilter: string;
  onSearchChange: (_value: string) => void;
  onStatusFilterChange: (_value: string) => void;
}

const GuestsFilters = ({ searchTerm, statusFilter, onSearchChange, onStatusFilterChange }: GuestsFiltersProps) => {
  return (
    <div className="flex gap-4 mb-6">
      <div className="relative flex-1">
        <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-4 w-4" />
        <Input placeholder="Search" value={searchTerm} onChange={(e) => onSearchChange(e.target.value)} className="pl-10" />
      </div>
      <Select value={statusFilter} onValueChange={onStatusFilterChange}>
        <SelectTrigger className="w-[180px]">
          <SelectValue placeholder="Status" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="all">All Status</SelectItem>
          <SelectItem value="BOOKED">Booked</SelectItem>
          <SelectItem value="COMPLETED">Completed</SelectItem>
          <SelectItem value="CANCELLED">Cancelled</SelectItem>
        </SelectContent>
      </Select>
    </div>
  );
};

export default GuestsFilters;
