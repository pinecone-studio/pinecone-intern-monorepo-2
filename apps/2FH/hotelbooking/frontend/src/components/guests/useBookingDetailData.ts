/* eslint-disable  */
import { useGetBookingQuery, useHotelQuery, useGetRoomQuery, useGetUserByIdQuery } from '@/generated';

interface User {
  firstName?: string;
  lastName?: string;
  email?: string;
}

interface Booking {
  userId?: string;
  roomId?: string;
}

const createGuestInfo = (user: User | undefined, booking: Booking | undefined) => ({
  firstName: user?.firstName || 'Guest',
  lastName: user?.lastName || booking?.userId?.slice(-8) || '',
  email: user?.email || `${booking?.userId?.slice(-8) || ''}@example.com`,
  phone: '+976 99112233', // Phone not available in user data
  guestRequest: 'No Request', // Guest request not available in user data
  roomNumber: `Room #${booking?.roomId?.slice(-3) || ''}`,
});

export const useBookingDetailData = (bookingId: string) => {
  const {
    data: bookingData,
    loading: bookingLoading,
    error: bookingError,
    refetch,
  } = useGetBookingQuery({
    variables: { getBookingId: bookingId },
    skip: !bookingId,
  });

  const { data: hotelData, loading: hotelLoading } = useHotelQuery({
    variables: { hotelId: bookingData?.getBooking?.hotelId || '' },
    skip: !bookingData?.getBooking?.hotelId,
  });

  const { data: roomData, loading: roomLoading } = useGetRoomQuery({
    variables: { getRoomId: bookingData?.getBooking?.roomId || '' },
    skip: !bookingData?.getBooking?.roomId,
  });

  const { data: userData, loading: userLoading } = useGetUserByIdQuery({
    variables: { input: { _id: bookingData?.getBooking?.userId || '' } },
    skip: !bookingData?.getBooking?.userId,
  });

  const booking = bookingData?.getBooking;
  const hotel = hotelData?.hotel;
  const room = roomData?.getRoom;
  const user = userData?.getUserById;

  // Use real guest info from user data
  const guestInfo = createGuestInfo(user as User, booking);

  const handleStatusUpdate = () => {
    refetch();
  };

  return {
    booking,
    hotel,
    room,
    guestInfo,
    loading: bookingLoading || hotelLoading || roomLoading || userLoading,
    error: bookingError,
    onStatusUpdate: handleStatusUpdate,
  };
};
