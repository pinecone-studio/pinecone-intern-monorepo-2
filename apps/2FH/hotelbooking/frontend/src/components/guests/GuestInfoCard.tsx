'use client';
import React from 'react';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import CheckoutModal from './CheckoutModal';

interface GuestInfoCardProps {
  guestInfo: {
    firstName: string;
    lastName: string;
    email: string;
    phone: string;
    guestRequest: string;
    roomNumber: string;
  };
  booking: {
    id: string;
    status: string;
    adults?: number | null;
    children?: number | null;
    checkInDate: any;
    checkOutDate: any;
  };
  onStatusUpdate: () => void;
}

const GuestInfoCard = ({ guestInfo, booking, onStatusUpdate }: GuestInfoCardProps) => {
  const formatDate = (date: any) => {
    const d = new Date(date);
    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    return {
      day: days[d.getDay()],
      month: months[d.getMonth()],
      date: d.getDate(),
      time: d.toLocaleTimeString('en-US', { hour: 'numeric', minute: '2-digit', hour12: true }),
    };
  };

  const getStatusBadgeStyle = (status: string) => {
    switch (status) {
      case 'BOOKED':
        return 'bg-blue-600 text-white';
      case 'COMPLETED':
        return 'bg-green-600 text-white';
      case 'CANCELLED':
        return 'bg-orange-500 text-white';
      default:
        return 'bg-blue-600 text-white';
    }
  };

  const checkInFormatted = formatDate(booking.checkInDate);
  const checkOutFormatted = formatDate(booking.checkOutDate);

  return (
    <Card>
      <CardHeader>
        <CardTitle className="text-lg font-semibold">Guest Info</CardTitle>
      </CardHeader>
      <CardContent>
        <div className="grid grid-cols-2 gap-4">
          <div>
            <label className="text-sm font-medium text-gray-600">First Name</label>
            <p className="text-gray-900">{guestInfo.firstName}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Last Name</label>
            <p className="text-gray-900">{guestInfo.lastName}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Status</label>
            <div className="mt-1">
              <Badge className={getStatusBadgeStyle(booking.status)}>{booking.status.charAt(0) + booking.status.slice(1).toLowerCase()}</Badge>
            </div>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Guests</label>
            <p className="text-gray-900">
              {(booking as any).adults || 1} adult{((booking as any).adults || 1) > 1 ? 's' : ''}, {(booking as any).children || 0} children
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Check in</label>
            <p className="text-gray-900">
              {checkInFormatted.month} {checkInFormatted.date}, {checkInFormatted.day}, {checkInFormatted.time}
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Check out</label>
            <p className="text-gray-900">
              {checkOutFormatted.month} {checkOutFormatted.date}, {checkOutFormatted.day}, {checkOutFormatted.time}
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Email</label>
            <p className="text-gray-900">{guestInfo.email}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Phone</label>
            <p className="text-gray-900">{guestInfo.phone}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Guest Request</label>
            <p className="text-gray-900">{guestInfo.guestRequest}</p>
          </div>
        </div>
        <div className="flex justify-end mt-6">
          <CheckoutModal bookingId={booking.id} currentStatus={booking.status} onStatusUpdate={onStatusUpdate} />
        </div>
      </CardContent>
    </Card>
  );
};

export default GuestInfoCard;
