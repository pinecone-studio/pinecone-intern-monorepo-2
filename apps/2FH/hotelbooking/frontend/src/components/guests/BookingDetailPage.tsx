/* eslint-disable  */
'use client';
import React from 'react';
import { Button } from '@/components/ui/button';
import { ArrowLeft } from 'lucide-react';
import { useRouter } from 'next/navigation';
import { useBookingDetailData } from './useBookingDetailData';
import GuestInfoCard from './GuestInfoCard';
import RoomDetailsCard from './RoomDetailsCard';
import PriceDetailCard from './PriceDetailCard';

interface BookingDetailPageProps {
  bookingId: string;
}

const BookingDetailPage = ({ bookingId }: BookingDetailPageProps) => {
  const router = useRouter();
  const { booking, room, guestInfo, loading, error, onStatusUpdate } = useBookingDetailData(bookingId);

  if (loading) {
    return (
      <div className="p-6">
        <div className="flex items-center justify-center h-64">
          <div className="text-gray-500">Loading booking details...</div>
        </div>
      </div>
    );
  }

  if (error || !booking) {
    return (
      <div className="p-6">
        <div className="flex items-center justify-center h-64">
          <div className="text-red-500">Error loading booking details</div>
        </div>
      </div>
    );
  }

  return (
    <div className="min-h-screen bg-gray-50 p-6">
      {/* Breadcrumb */}
      <div className="mb-6">
        <nav className="text-sm text-gray-600">
          <span className="hover:text-gray-900 cursor-pointer">Guests</span>
          <span className="mx-2">&gt;</span>
          <span className="text-gray-900">Guest Info</span>
        </nav>
      </div>

      {/* Header */}
      <div className="flex items-center gap-4 mb-6">
        <Button variant="ghost" size="sm" onClick={() => router.back()} className="p-2" aria-label="Go back">
          <ArrowLeft className="h-4 w-4" />
        </Button>
        <h1 className="text-2xl font-bold text-gray-900">
          {guestInfo.firstName} {guestInfo.lastName}
        </h1>
      </div>

      <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
        {/* Left Panel - Guest Information */}
        <div className="space-y-6">
          <GuestInfoCard
            guestInfo={guestInfo}
            booking={{
              id: booking.id,
              status: booking.status || 'BOOKED',
              adults: booking.adults,
              children: booking.children,
              checkInDate: booking.checkInDate,
              checkOutDate: booking.checkOutDate,
            }}
            onStatusUpdate={onStatusUpdate}
          />
        </div>

        {/* Right Panel - Room and Price Details */}
        <div className="space-y-6">
          <RoomDetailsCard room={room} booking={booking} />
          <PriceDetailCard room={room} />
        </div>
      </div>
    </div>
  );
};

export default BookingDetailPage;
