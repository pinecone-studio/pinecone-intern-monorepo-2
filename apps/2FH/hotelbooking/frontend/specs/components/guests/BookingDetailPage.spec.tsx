/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BookingDetailPage from '@/components/guests/BookingDetailPage';
import { useBookingDetailData } from '@/components/guests/useBookingDetailData';
import { useRouter } from 'next/navigation';

// Mock dependencies
jest.mock('@/components/guests/useBookingDetailData', () => ({
  useBookingDetailData: jest.fn(),
}));

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

jest.mock('@/components/guests/GuestInfoCard', () => ({
  __esModule: true,
  default: ({ guestInfo, booking, onStatusUpdate }: any) => (
    <div data-testid="guest-info-card">
      <div>
        Guest: {guestInfo.firstName} {guestInfo.lastName}
      </div>
      <div>Status: {booking.status}</div>
      <button onClick={onStatusUpdate}>Update Status</button>
    </div>
  ),
}));

jest.mock('@/components/guests/RoomDetailsCard', () => ({
  __esModule: true,
  default: ({ room, booking }: any) => (
    <div data-testid="room-details-card">
      <div>Room: {room?.name || 'Unknown Room'}</div>
      <div>Room ID: {booking.roomId}</div>
    </div>
  ),
}));

jest.mock('@/components/guests/PriceDetailCard', () => ({
  __esModule: true,
  default: ({ room }: any) => (
    <div data-testid="price-detail-card">
      <div>Price: {room?.pricePerNight || 0}₮</div>
    </div>
  ),
}));

const mockUseBookingDetailData = useBookingDetailData as jest.MockedFunction<typeof useBookingDetailData>;
const mockUseRouter = useRouter as jest.MockedFunction<typeof useRouter>;

describe('BookingDetailPage', () => {
  const mockBooking = {
    id: 'booking-1',
    userId: 'user-1',
    roomId: 'room-1',
    hotelId: 'hotel-1',
    status: 'BOOKED',
    adults: 2,
    children: 1,
    checkInDate: '2024-02-01',
    checkOutDate: '2024-02-03',
  };

  const mockRoom = {
    id: 'room-1',
    name: 'Deluxe Suite',
    pricePerNight: 150000,
  };

  const mockGuestInfo = {
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@example.com',
    phone: '+976 99112233',
    guestRequest: 'No Request',
    roomNumber: 'Room #001',
  };

  const mockRouter = {
    back: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    refresh: jest.fn(),
    forward: jest.fn(),
    prefetch: jest.fn(),
  };

  const defaultProps = {
    bookingId: 'booking-1',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseRouter.mockReturnValue(mockRouter);
  });

  it('renders loading state', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: undefined,
      room: undefined,
      guestInfo: mockGuestInfo,
      loading: true,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Loading booking details...')).toBeInTheDocument();
  });

  it('renders error state', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: undefined,
      room: undefined,
      guestInfo: mockGuestInfo,
      loading: false,
      error: new Error('Failed to load booking'),
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Error loading booking details')).toBeInTheDocument();
  });

  it('renders error state when booking is null', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: null,
      room: undefined,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Error loading booking details')).toBeInTheDocument();
  });

  it('renders booking details when data is available', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Guests')).toBeInTheDocument();
    expect(screen.getByText('Guest Info')).toBeInTheDocument();
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByTestId('guest-info-card')).toBeInTheDocument();
    expect(screen.getByTestId('room-details-card')).toBeInTheDocument();
    expect(screen.getByTestId('price-detail-card')).toBeInTheDocument();
  });

  it('displays breadcrumb navigation', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Guests')).toBeInTheDocument();
    expect(screen.getByText('Guest Info')).toBeInTheDocument();
  });

  it('calls router.back when back button is clicked', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    const backButton = screen.getByRole('button', { name: 'Go back' });
    fireEvent.click(backButton);

    expect(mockRouter.back).toHaveBeenCalled();
  });

  it('renders with correct page structure', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    const pageContainer = screen.getByTestId('guest-info-card').closest('.min-h-screen');
    expect(pageContainer).toHaveClass('bg-gray-50', 'p-6');
  });

  it('renders grid layout correctly', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    const gridContainer = document.querySelector('.grid.grid-cols-1.lg\\:grid-cols-2');
    expect(gridContainer).toBeInTheDocument();
  });

  it('passes correct props to child components', () => {
    const onStatusUpdate = jest.fn();
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate,
    });

    render(<BookingDetailPage {...defaultProps} />);

    // Check that guest info is passed correctly
    expect(screen.getByText('Guest: John Doe')).toBeInTheDocument();
    expect(screen.getByText('Status: BOOKED')).toBeInTheDocument();

    // Check that room details are passed correctly
    expect(screen.getByText('Room: Deluxe Suite')).toBeInTheDocument();
    expect(screen.getByText('Room ID: room-1')).toBeInTheDocument();

    // Check that price details are passed correctly
    expect(screen.getByText('Price: 150000₮')).toBeInTheDocument();
  });

  it('handles status update', () => {
    const onStatusUpdate = jest.fn();
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate,
    });

    render(<BookingDetailPage {...defaultProps} />);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    expect(onStatusUpdate).toHaveBeenCalled();
  });

  it('handles missing room data', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: null,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Room: Unknown Room')).toBeInTheDocument();
    expect(screen.getByText('Price: 0₮')).toBeInTheDocument();
  });

  it('handles missing guest info', () => {
    const incompleteGuestInfo = {
      firstName: 'John',
      lastName: '',
      email: '',
      phone: '',
      guestRequest: '',
      roomNumber: '',
    };

    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: incompleteGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(screen.getByText('Guest: John')).toBeInTheDocument();
  });

  it('calls useBookingDetailData with correct booking ID', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    expect(mockUseBookingDetailData).toHaveBeenCalledWith('booking-1');
  });

  it('matches snapshot for loading state', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: undefined,
      room: undefined,
      guestInfo: mockGuestInfo,
      loading: true,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    const { container } = render(<BookingDetailPage {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for loaded state', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: mockBooking,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    const { container } = render(<BookingDetailPage {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for error state', () => {
    mockUseBookingDetailData.mockReturnValue({
      booking: undefined,
      room: undefined,
      guestInfo: mockGuestInfo,
      loading: false,
      error: new Error('Failed to load booking'),
      onStatusUpdate: jest.fn(),
    });

    const { container } = render(<BookingDetailPage {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('uses fallback status when booking status is undefined', () => {
    const bookingWithUndefinedStatus = {
      ...mockBooking,
      status: undefined,
    };

    mockUseBookingDetailData.mockReturnValue({
      booking: bookingWithUndefinedStatus,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    // The GuestInfoCard mock should receive the fallback status 'BOOKED'
    expect(screen.getByText('Status: BOOKED')).toBeInTheDocument();
  });

  it('uses fallback status when booking status is null', () => {
    const bookingWithNullStatus = {
      ...mockBooking,
      status: null,
    };

    mockUseBookingDetailData.mockReturnValue({
      booking: bookingWithNullStatus,
      room: mockRoom,
      guestInfo: mockGuestInfo,
      loading: false,
      error: undefined,
      onStatusUpdate: jest.fn(),
    });

    render(<BookingDetailPage {...defaultProps} />);

    // The GuestInfoCard mock should receive the fallback status 'BOOKED'
    expect(screen.getByText('Status: BOOKED')).toBeInTheDocument();
  });
});
