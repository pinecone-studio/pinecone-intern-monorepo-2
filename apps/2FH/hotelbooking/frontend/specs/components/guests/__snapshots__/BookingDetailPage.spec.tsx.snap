// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDetailPage matches snapshot for error state 1`] = `
<div
  class="p-6"
>
  <div
    class="flex items-center justify-center h-64"
  >
    <div
      class="text-red-500"
    >
      Error loading booking details
    </div>
  </div>
</div>
`;

exports[`BookingDetailPage matches snapshot for loaded state 1`] = `
<div
  class="min-h-screen bg-gray-50 p-6"
>
  <div
    class="mb-6"
  >
    <nav
      class="text-sm text-gray-600"
    >
      <span
        class="hover:text-gray-900 cursor-pointer"
      >
        Guests
      </span>
      <span
        class="mx-2"
      >
        &gt;
      </span>
      <span
        class="text-gray-900"
      >
        Guest Info
      </span>
    </nav>
  </div>
  <div
    class="flex items-center gap-4 mb-6"
  >
    <button
      aria-label="Go back"
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md p-2"
    >
      <svg
        class="lucide lucide-arrow-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m12 19-7-7 7-7"
        />
        <path
          d="M19 12H5"
        />
      </svg>
    </button>
    <h1
      class="text-2xl font-bold text-gray-900"
    >
      John
       
      Doe
    </h1>
  </div>
  <div
    class="grid grid-cols-1 lg:grid-cols-2 gap-6"
  >
    <div
      class="space-y-6"
    >
      <div
        data-testid="guest-info-card"
      >
        <div>
          Guest: 
          John
           
          Doe
        </div>
        <div>
          Status: 
          BOOKED
        </div>
        <button>
          Update Status
        </button>
      </div>
    </div>
    <div
      class="space-y-6"
    >
      <div
        data-testid="room-details-card"
      >
        <div>
          Room: 
          Deluxe Suite
        </div>
        <div>
          Room ID: 
          room-1
        </div>
      </div>
      <div
        data-testid="price-detail-card"
      >
        <div>
          Price: 
          150000
          â‚®
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BookingDetailPage matches snapshot for loading state 1`] = `
<div
  class="p-6"
>
  <div
    class="flex items-center justify-center h-64"
  >
    <div
      class="text-gray-500"
    >
      Loading booking details...
    </div>
  </div>
</div>
`;
